Intel SGX SDK Porting Guide (Linux) (#3542)
openenclave/openenclave: docs/GettingStartedDocs/IntelSDKPortingGuideLinux.md

> # Intel SGX SDK to Open Enclave SDK Porting Guide (Linux)

Reviewer: Since the project was submitted to the CCC, the name has been listed in CCC docs as "SGX SDK for Linux" which is how it appears at http://confidentialcomputing.io (click on "Projects"), rather than "Intel SGX SDK".   Clicking on that link takes you to an Intel page that still calls it the "Intel SGX SDK for Linux" (but notice it still has "for Linux" in the SDK name).

Since it hasn't completed the CCC acceptance process yet, it's up to you whether you prefix it with "Intel" (which implies you want to advertise that it's NOT yet a full CCC project) or remove the "Intel" (which implies it IS a CCC project like the CCC home page implies), but all references in this doc to the name should be consistent in terms of keeping the "... for Linux" suffix in the SDK name.

# Intel SGX SDK for Linux to Open Enclave SDK Porting Guide

> Intel SGX SDK for Linux and Open Enclave SDK share many design principles but

Reviewer: ```suggestion
The [Intel SGX SDK for Linux](https://01.org/intel-software-guard-extensions) and the Open Enclave SDK share many design principles but
```
Besides adding a link, note the grammar fix to add "the".  (By comparison the linked page always uses "the", as do most other OE SDK docs, including lines 282-284 of this PR).

differ in implementations. It requires source code changes to build enclaves

> with Open Enclave SDK that were developed initially with Intel SGX SDK for

Reviewer: ```suggestion
with the Open Enclave SDK that were developed initially with the Intel SGX SDK for
```

Linux.

> Please note that this doc focuses on **Linux**, and applies to *Intel SGX SDK

Reviewer: ```suggestion
Please note that this doc focuses on **Linux**, and applies to the *Intel SGX SDK
```

for Linux* only unless explicitly noted otherwise.

## Create Project Build Files

> Open Enclave SDK supports a number of build systems, among which `cmake` is a

Reviewer: ```suggestion
The Open Enclave SDK supports a number of build systems, among which `cmake` is a
```

convenient one. This section shows how to build an enclave using `cmake`. For
those not familiar with `cmake`, a tutorial is available at
https://cmake.org/cmake/help/latest/guide/tutorial/index.html.

Firstly, a `CMakeLists.txt` file needs to be created in the enclave's source
directory. Open Enclave requires `cmake 3.11` or later so the first statement
should be:

```cmake
cmake_minimum_required(VERSION 3.11)
```

Then an enclave is just a C/C++ project.

```cmake
project("MyEnclaveProject" LANGUAGE C CXX)
```

Next, import the `OpenEnclave` package. Please note that the statement below

> requires appending */path/to/OpenEnclave_InstallDir* to `$CMAKE_PREFIX_PATH`

Author: Updated the doc. Please review the latest commit.

```cmake
find_package(OpenEnclave CONFIG REQUIRED)
```

Now generate the ECall/OCall bridge/proxy routines for the enclave.

```cmake
add_custom_command(
  OUTPUT MyEnclave_t.h MyEnclave_t.c MyEnclave_args.h
  DEPENDS ${CMAKE_SOURCE_DIR}/MyEnclave.edl
  COMMAND openenclave::oeedger8r --trusted ${CMAKE_SOURCE_DIR}/MyEnclave.edl
          --search-path ${OE_INCLUDEDIR})
```

Finally, build the enclave as an executable. Please don't forget to include
`MyEnclave_t.c` (generated above) in the source list.

```cmake
add_executable(MyEnclave MyEnclave.cpp ${CMAKE_CURRENT_BINARY_DIR}/MyEnclave_t.c)

# Current API version
target_compile_definitions(MyEnclave PUBLIC OE_API_VERSION=2)

# Needed by the generated MyEnclave_t.c
target_include_directories(MyEnclave PRIVATE ${CMAKE_CURRENT_BINARY_DIR})

target_link_libraries(MyEnclave openenclave::oeenclave openenclave::oelibc)
```

Enclaves must be signed before they can be loaded. How to sign an enclave
depends on how the signing private key is managed in your project. For
demonstration purposes, the following snippet generates a random RSA key pair
to sign the enclave. An enclave configuration file (`MyEnclave.conf` below) may

> also be provided. Details of enclave configuration/settings are described later
> in this document.

Reviewer: ```suggestion
also be provided. Details of enclave configuration/settings are described
[later in this document](#migrate-enclave-settings).
```

```cmake
# Generate key
add_custom_command(
  OUTPUT private.pem public.pem
  COMMAND openssl genrsa -out private.pem -3 3072
  COMMAND openssl rsa -in private.pem -pubout -out public.pem)

# Sign enclave
add_custom_command(
  OUTPUT MyEnclave.signed
  DEPENDS MyEnclave MyEnclave.conf private.pem
  COMMAND openenclave::oesign sign -e $<TARGET_FILE:MyEnclave>
          -c ${CMAKE_SOURCE_DIR}/MyEnclave.conf -k private.pem)
```

Readers are encouraged to look for complete examples under
*/path/to/OpenEnclave_InstallDir*/share/openenclave/samples directory.

## Migrate Enclave Settings

**Enclave Settings**, also known as **Enclave Metadata**, refers to information
consumed by enclave loaders to instantiate enclaves, such as heap size, stack
size, number of trusted hardware threads (i.e., number of TCS's), etc.

Enclave settings are specified in human readable text formats by enclave

> developers, referred to as *configuration files*. Both Intel SGX SDK for Linux
> and Open Enclave SDK provide tools to compile configuration files into their

Reviewer: ```suggestion
developers, referred to as *configuration files*. Both the Intel SGX SDK for Linux
and the Open Enclave SDK provide tools to compile configuration files into their
```

binary form and embed them into the final enclave image. However, they differ
in both format and feature set.

### Configuration File Formats

> Intel SGX SDK for Linux adopted an XML format for encoding enclave settings in

Reviewer: ```suggestion
The Intel SGX SDK for Linux adopted an XML format for encoding enclave settings in
```

text form, which is usually named as *Enclave*.config.xml. The signing tool
(i.e., `sgx_sign`) converts it into binary form and stores it in a dedicated
section (i.e., `.sgxmeta`) of the enclave's ELF image before it calculates the
enclave's measurement (i.e., `SIGSTRUCT::MRENCLAVE`). At runtime, the ELF
section `.sgxmeta` is consumed by the enclave loader to instantiate the exact
enclave that matches the measurement calculated by the signing tool.

> Below comes from the sample code - SampleEnclave, of Intel SGX SDK for Linux.

Reviewer: ```suggestion
Below comes from the sample code - SampleEnclave, of the Intel SGX SDK for Linux.
```

```xml
<EnclaveConfiguration>
  <ProdID>0</ProdID>
  <ISVSVN>0</ISVSVN>
  <StackMaxSize>0x40000</StackMaxSize>
  <HeapMaxSize>0x100000</HeapMaxSize>
  <TCSNum>10</TCSNum>
  <TCSPolicy>1</TCSPolicy>
  <!-- Recommend changing 'DisableDebug' to 1 to make the enclave undebuggable for enclave release -->
  <DisableDebug>0</DisableDebug>
  <MiscSelect>0</MiscSelect>
  <MiscMask>0xFFFFFFFF</MiscMask>
</EnclaveConfiguration>
```

> Rather than XML, Open Enclave SDK uses plaintext files instead. A configuration

Reviewer: ```suggestion
Rather than XML, the Open Enclave SDK uses plaintext files instead. A configuration
```

file, usually named *enclave*.conf, is supplied to the signing tool (i.e.,
`oesign`) command line to govern the instantiation of the enclave. The compiled
metadata is then stored in a dedicated ELF section named `.oeinfo`.
Additionally, Open Enclave SDK provides `OE_SET_ENCLAVE_SGX`, a C macro for
embedding default enclave settings in C source files. *enclave*.conf is in fact
optional and is necessary only if some of those defaults provided to
`OE_SET_ENCLAVE_SGX` macro need overridden. Under the hood,
`OE_SET_ENCLAVE_SGX` is expanded to instantiation of an
`oe_sgx_enclave_properties_t` structure in `.oeinfo` section. Detailed
information can be found in
[buildandsign.md](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/buildandsign.md)
in Open Enclave SDK documentation.

> Below is the same configuration as above but in Open Enclave SDK's

Reviewer: ```suggestion
Below is the same configuration as above but in the Open Enclave SDK's
```

*enclave*.conf format.

```
# <ProdID>0</ProdID>
ProductID=0

# <ISVSVN>0</ISVSVN>
SecurityVersion=0

# <StackMaxSize>0x40000</StackMaxSize>
NumStackPages=64

# <HeapMaxSize>0x100000</HeapMaxSize>
NumHeapPages=256

# <TCSNum>10</TCSNum>
NumTCS=10

# <DisableDebug>0</DisableDebug>
Debug=1

# There are no equivalent Open Enclave enclave settings for the following
# <TCSPolicy>1</TCSPolicy>
# <MiscSelect>0</MiscSelect>
# <MiscMask>0xFFFFFFFF</MiscMask>
```

### Supported Enclave Settings by Intel SGX SDK for Linux and Open Enclave SDK

> At the time of this writing, Intel SGX SDK for Linux supports a superset of

Reviewer: ```suggestion
At the time of this writing, the Intel SGX SDK for Linux supports a superset of the
```

Open Enclave SDK's features, hence not every element of Intel's

> *Enclave*.conf.xml has a corresponding setting in Open Enclave SDK's

Reviewer: ```suggestion
*Enclave*.conf.xml has a corresponding setting in the Open Enclave SDK's
```

|`<ReleaseType>`||`bool`|`1` indicates a release build|Intel SDK copies this bit to MSB of `SIGSTRUCT::HEADER`. Open Enclave does *NOT* support configuring this bit but hard-codes it to `0`.This bit is *NOT* documented in [SDM](https://software.intel.com/content/www/us/en/develop/articles/intel-sdm.html). User enclaves shall avoid using this bit.
|`<IntelSigned>`||`bool`|If `1`, set `SIGSTRUCT::VENDOR` to `0x8086` (or `0` otherwise)|Open Enclave does *NOT* support configuring this field currently but hard-codes it to `0`. Per SDM, this field is informational.
|`<ProvisionKey>`||`bool`|`1` to grant access to *Provision Key*. This corresponds to bit 4 of `SIGSTRUCT::ATTRIBUTES`|`Debug` is the only attribute configurable via Open Enclave's .conf file. All other attributes can only be configured by enclosing an `oe_sgx_enclave_properties_t` structure manually in the `.oeinfo` section in a source file.|`<LaunchKey>`||`bool`|`1` to grant access to *Launch Key*. This corresponds to bit 5 of `SIGSTRUCT::ATTRIBUTES`|Similar to `<ProvisionKey>` above, manual instantiation of `oe_sgx_enclave_properties_t` is required.

> |`<DisableDebug>`|`Debug`|`bool`|Indicate whether debugging is allowed|Intel and Open Enclave SDKs use different polarity - i.e. `<DisableDebug>1</DisableDebug>` is equivlanet to `Debug=0`.

Reviewer: ```suggestion
|`<DisableDebug>`|`Debug`|`bool`|Indicate whether debugging is allowed|The Intel SGX SDK for Linux and the Open Enclave SDKs use different polarity - i.e., `<DisableDebug>1</DisableDebug>` is equivalent to `Debug=0`.
```

|`<HW>`||`uint32_t`|Hardware verions. This occupies the space of `SIGSTRUCT::SWDEFINED`|Currently it's used only by Intel's LE (Launch Enclave). Open Enclave does *NOT* support configuring this field currently but hard-codes it to `0`. User enclaves shall avoid using it.
|`<TCSNum>`|`NumTCS`|`uint32_t`|Number of TCS's (trusted threads)|This is the number of TCS's, and is also the initial number of TCS's on SGX v2. Open Enclave supports only SGX v1 at the moment.
|`<TCSMaxNum>`||`uint32_t`|Maximal number of TCS's|TCS's can be added at runtime on SGX v2. Open Enclave supports only SGX v1 at the moment.
|`<TCSMinPool>`||`uint32_t`|Minimal number of TCS's to keep|TCS's can be removed at runtime on SGX v2. Open Enclave supports only SGX v1 at the moment.

> |`<TCSPolicy>`||`bool`|`0` to bind TCS to untrusted thread, `1` to unbind them|Open Enclave SDK never does not bind TCS's to untrusted threads.

Reviewer: I can't parse "never does not bind".
```suggestion
|`<TCSPolicy>`||`bool`|`0` to bind TCS to untrusted thread, `1` to unbind them|Open Enclave SDK never binds TCS's to untrusted threads.
```

|`<ISVEXTPRODID_L>`||`uint64_t`|See above|

As mentioned in the *Notes* column above, certain missing feature, such as
those controlling SGX enclave attribute bits (e.g., `<LaunchKey>`,
`<ProvisionKey>`), could still be enabled by setting
`oe_sgx_enclave_properties_t::config.attributes` manually, even though they
aren't supported explicitly by the `OE_SET_ENCLAVE_SGX` macro or *enclave*.conf
file. Below is an example that grants the enclave access to `PROVISION_KEY`.

```C
OE_INFO_SECTION_BEGIN
volatile const oe_sgx_enclave_properties_t oe_enclave_properties_sgx = {
    .header = {.size = sizeof(oe_sgx_enclave_properties_t),
               .enclave_type = OE_ENCLAVE_TYPE_SGX,
               .size_settings = {.num_heap_pages = 512,
                                 .num_stack_pages = 512,
                                 .num_tcs = 4}},
    .config = {.product_id = 1234,
               .security_version = 5678,
               .attributes = OE_SGX_FLAGS_PROVISION_KEY |
                             OE_MAKE_ATTRIBUTES(0)},
    .end_marker = 0xecececececececec};
OE_INFO_SECTION_END
```

If compatibilities with both SDKs are desired, avoid using features specific to
either SDK.

> As a final note, neither Intel SGX SDK for Linux nor Open Enclave SDK provides

Reviewer: ```suggestion
As a final note, neither the Intel SGX SDK for Linux nor the Open Enclave SDK provides
```

configuration settings for enabling/disabling X features (e.g., AVX, AVX-512,
etc.) explicitly. Open Enclave's SGX enclave loader uses the enabled X features
on the local platform to initialize `SECS::ATTRIBUTES::XFRM`, and hard-codes
`SIGSTRUCT::ATTRIBUTEMASK::XFRM` to `0`. That is, X features are *NOT* enforced
and must *NOT* be relied upon for security.

## Migrate ECall/OCall Definitions (EDL Files)

> Both Intel SGX SDK for Linux and Open Enclave SDK support the same grammer for

Reviewer: ```suggestion
Both the Intel SGX SDK for Linux and the Open Enclave SDK support the same grammar for
```

defining trusted/untrusted functions (aka. ECalls/OCalls) in EDL files.
However, built-in OCalls are defined in different headers. Intel's built-in
OCalls are defined in `sgx_tstdc.edl` while Open Enclave SDK's are defined in
`platform.edl`.

Most EDL files include (by `include` statements) common C headers for both host
and enclave sides. The most commonly included header is the one defining SGX

> architectural structures, which is `arch.h` in Intel SGX SDK for Linux or
> `openenclave/bits/sgx/sgxtypes.h` in Open Enclave SDK. Please also note that

Reviewer: ```suggestion
architectural structures, which is `arch.h` in the Intel SGX SDK for Linux or
`openenclave/bits/sgx/sgxtypes.h` in the Open Enclave SDK. Please also note that
```

some structures may be named differently, e.g., the EINITTOKEN architectural

> structure is defined as `token_t` in Intel SGX SDK for Linux but `einittoken_t`
> in Open Enclave SDK.

Reviewer: ```suggestion
structure is defined as `token_t` in the Intel SGX SDK for Linux but `einittoken_t`
in the Open Enclave SDK.
```

The code snippet below shows a way to include/import C headers and EDL
definitions conditionally, in order to be compatible with both SDKs.

```
enclave {
#ifdef OEEDGER8R
    include "openenclave/bits/sgx/sgxtypes.h";
    from "openenclave/edl/sgx/platform.edl" import *;
#else
    include "arch.h";
    from "sgx_tstdc.edl" import *;
#endif
    /* ECall/OCall definitions go here */
}
```

> The macro `OEEDGER8R` used above could be defined by appending `-DOEEDGER8R` to
> `oeedger8r` command line. Please note that only `oeedger8r` supports defining
> macros at command line, even though both `sgx_edger8r` and `oeedger8r` support
> preprocessing.

Reviewer 2: I would suggest being explicit and offering examples of both here:

````suggestion
For example, if using `oeedger8r`:

```bash
oeedger8r --trusted ${CMAKE_SOURCE_DIR}/MyEnclave.edl --search-path ${OE_INCLUDEDIR}) -DOEEDGER8R
```

`sgx_edger8r` also supports macro preprocessing in EDL files, but does not accept macro definitions as arguments. To use `sgx_edger8r`:
````

... you'll need to fill in an equivalent for sgx_edger8r as I'm unfamiliar with the syntax off the top of my head.

The last thing worth noting is that Open Enclave doesn't support nested ECall
(i.e., an ECall in the context of an OCall) like the Intel SGX SDK for Linux
does. Existing enclaves making use of nested ECalls need to be reworked to be
compatible with the Open Enclave SDK.

## Port C/C++ Source Code

> Given similarities in the architectures of both SDKs, there shall not be any

Reviewer 2: Using the present tense "shall" here indicates an imperative which seems unintended here. 
```suggestion
Given similarities in the architectures of both SDKs, there should not be any
```

significant code flow/logic changes required. However, source code

> incompatibilities still exist in

Reviewer 2: ```suggestion
incompatibilities still exist in:
```

- Header files - They are structured and/or named differently. Fortunately,
  Open Enclave provides 2 comprehensive headers, namely `openenclave/enclave.h`
  and `openenclave/host.h`, to be included by trusted and untrusted code,
  respectively. A single `#include` should suffice in most cases.
- APIs - Most Open Enclave APIs are prefixed by `oe_` while Intel's APIs are by
  `sgx_`. Moreover, some APIs may take parameters in different orders.
- Structure definitions - Structure members may be named differently. Some

>   structures are organized differently. For example, Intel SGX SDK for Linux

Reviewer: ```suggestion
  structures are organized differently. For example, the Intel SGX SDK for Linux
```

  defines EINITTOKEN as `token_t` with all MAC'ed fields captured in a child
  structure `launch_body_t`; while Open Enclave defines it as a flat
  `einittoken_t` structure.
- Crypto lib - Intel SGX SDK for Linux supports 2 crypto libs - IPP and

>   OpenSSL, and provides a wrapper layer to unify crypto APIs. Open Enclave

Reviewer 3: Thanks a lot for writing this up Cedric!
I think this is a good start.
Will this be expanded to include
1) What customers would do if using the key exchange lib
2) What about features such as sealing
3) Subtle differences in edl support, switchless support, threadbinding , exception handling etc.

Author: Thanks and all those are good questions!

In fact, we (Intel) had an internal meeting and we decide to put in more contents/details regarding Intel SGX SDK features that are implemented differently or unimplemented by OE SDK. Just look out for updates to this PR.

Then this doc will become a *real* good start point to be merged into `master`. We will then contribute to OE SDK with new features deemed necessary per customer feedbacks and update this doc accordingly going forward.

Reviewer 3: Thanks! Would the expectation be that you would merge this PR and open new PRs with more details?

Author: We are preparing more contents for the last section of the document, such as difference in core APIs, libraries and signing tool, We are ok to either extend this PR or submit new PRs. Merging it right now will make it available for broader review, which is our preference. I'll incorporate the remaining comments and make it ready for merge shortly.

>   supports only MbedTLS and provides no wrapper.

Reviewer 2: ```suggestion
only supports enclave applications calling [MbedTLS](/docs/MbedtlsSupport.md)
directly and not through an SDK wrapper.
```

> Generally, there's no good way to fast port source code rather than reacting to

Reviewer 4: ```suggestion
Generally, there's no good way to fast port source code other than reacting to
```

> Generally, there's no good way to fast port source code rather than reacting to

Reviewer 2: I would suggest dropping this sentence as it doesn't provide any helpful information to a reader.

Author: Agreed. Removed.
